ENH: add np.stack

shoyer opened the pull request:
The motivation here is to present a uniform and N-dimensional interface for joining arrays along a new axis, similarly to how `concatenate` provides a uniform and N-dimensional interface for joining arrays along an existing axis.
## Background

Currently, users can choose between `hstack`, `vstack`, `column_stack` and `dstack`, but none of these functions handle N-dimensional input. In my opinion, it's also difficult to keep track of the differences between these methods and to predict how they will handle input with different dimensions.

In the past, my preferred approach has been to either construct the result array explicitly and use indexing for assignment, to or use `np.array` to stack along the first dimension and then use `transpose` (or a similar method) to reorder dimensions if necessary. This is pretty awkward.

I brought this proposal up a few weeks on the numpy-discussion list:
http://mail.scipy.org/pipermail/numpy-discussion/2015-February/072199.html

I also received positive feedback on Twitter:
https://twitter.com/shoyer/status/565937244599377920
## Implementation notes

The one line summaries for `concatenate` and `stack` have been (re)written to mirror each other, and to make clear that the distinction between these functions is whether they join over an existing or new axis.

In general, I've tweaked the documentation and docstrings with an eye toward pointing users to `concatenate`/`stack`/`split` as a fundamental set of basic array manipulation routines, and away from `array_split`/`{h,v,d}split`/`{h,v,d,column_}stack`

I put this implementation in `numpy.core.shape_base` alongside `hstack`/`vstack`, but it appears that there is also a `numpy.lib.shape_base` module that contains another larger set of functions, including `dstack`. I'm not really sure where this belongs (or if it even matters).

Finally, it might be a good idea to write a masked array version of `stack`. But I don't use masked arrays, so I'm not well motivated to do that.

shoyer committed:
ENH: add np.stack

The motivation here is to present a uniform and N-dimensional interface for
joining arrays along a new axis, similarly to how `concatenate` provides a
uniform and N-dimensional interface for joining arrays along an existing axis.

Background
~~~~~~~~~~

Currently, users can choose between `hstack`, `vstack`, `column_stack` and
`dstack`, but none of these functions handle N-dimensional input. In my
opinion, it's also difficult to keep track of the differences between these
methods and to predict how they will handle input with different
dimensions.

In the past, my preferred approach has been to either construct the result
array explicitly and use indexing for assignment, to or use `np.array` to
stack along the first dimension and then use `transpose` (or a similar method)
to reorder dimensions if necessary. This is pretty awkward.

I brought this proposal up a few weeks on the numpy-discussion list:
http://mail.scipy.org/pipermail/numpy-discussion/2015-February/072199.html

I also received positive feedback on Twitter:
https://twitter.com/shoyer/status/565937244599377920

Implementation notes
~~~~~~~~~~~~~~~~~~~~

The one line summaries for `concatenate` and `stack` have been (re)written to
mirror each other, and to make clear that the distinction between these functions
is whether they join over an existing or new axis.

In general, I've tweaked the documentation and docstrings with an eye toward
pointing users to `concatenate`/`stack`/`split` as a fundamental set of basic
array manipulation routines, and away from
`array_split`/`{h,v,d}split`/`{h,v,d,column_}stack`

I put this implementation in `numpy.core.shape_base` alongside `hstack`/`vstack`,
but it appears that there is also a `numpy.lib.shape_base` module that contains
another larger set of functions, including `dstack`. I'm not really sure where
this belongs (or if it even matters).

Finally, it might be a good idea to write a masked array version of `stack`.
But I don't use masked arrays, so I'm not well motivated to do that.

seberg commented on numpy/core/shape_base.py:
```diff
@@ -275,3 +278,62 @@ def hstack(tup):
+    if axis < 0:
+        axis += result_ndim
+    sl = (slice(None),) * axis + (_nx.newaxis,)
+    return _nx.concatenate([arr[sl] for arr in arrays], axis=axis)
```
I don't trust this on first sight (but if you have a test, ignore the comment). What if the second array has more dimensions then the first one?
One suggestion for the API. How about adding an `ndmin` kwarg. That way hstack, etc. would really be just special cases of this function, I believe (though whether we should implement them as such, I don't know).

mhvk replied:
To me at least, it would be clearer if one were to do: 
```
new_shape = array[0].shape[:axis] + [_nx.newaxis] + array[0].shape[axis:]
return _nx.concatenate([arr.reshape(new_shape) for arr in arrays], axis=axis)
```

(note that with the broadcasting suggestion, it might be faster to use the machinery from #5371 to get the broadcast shape at the top, and then use `broadcast_to(arr, new_shape)` here. Alternatively, use `b = np.broadcast(*arrays)` above, and `new_shape = b.shape[:axis] ...` here).

shoyer replied:
@seberg  I find the `ndmin` behavior of the `{v,h,d}stack` functions confusing. I would say that the lack of automated rules for adding dimensions is a feature of this function :).

@mhvk My only hesitation with using reshape is that I know it can result in copies instead of views in some edge cases. So, I usually stick to indexing to be safe.

mhvk replied:
@shoyer - actually, looking better both your method and mine can lead to unexpected errors when a later array has fewer dimensions than the first one, like::
```
np.stack([np.arange(9.).reshape(3,3), np.arange(3)], axis=2)
```
At least, I would think you would get `sl=[:,:,np.newaxis]` in which case you get an `IndexError: too many indices` on the second array (rather than the much more informative `ValueError` that `concatenate` would throw).

Though it would perhaps be needlessly expensive to test for this beforehand. Maybe just
```
try:
    return _nx.concatenate(...)
except IndexError:
    raise ValueError(<like concatenate>)
```

shoyer replied:
Good catch -- fixed. Though again, weird stuff will happen if you pass an array whose dimensions can't be expanded in here. I would think that would include np.matrix, but look at what it does:
```
In [8]: np.mat('1')[:, :, None].shape
Out[8]: (1, 1, 1)
```

mhvk replied:
To make the indexing slightly safer, one could explicitly use `reshape(arr.shape[:axis] + [np.newaxis] + arr.shape[axis:])` after all -- in that case one uses both the `shape` and `reshape` of the array subclass itself. Though, `matrix` is again interesting...
```
np.mat([[1,2],[3,4]]).reshape(2, 1, 2).shape
# (2, 2)
```
I don't know what's worse... Though again I don't feel one should "punish" other subclasses for strange behaviour of some.

mhvk commented on numpy/core/shape_base.py:
```diff
@@ -275,3 +278,62 @@ def hstack(tup):
+           [3, 4]])
+
+    """
+    arrays = [_nx.asarray(arr) for arr in arrays]
```
Would this be more general if one first broadcast the arrays against each other? One could simply use
```
arrays = _nx.broadcast_arrays(*arrays)
```

mhvk replied:
As one of astropy's `Quantity` maintainer, I try to encourage the use of `asanyarray` rather than `asarray` -- don't assume that every `ndarray` subclass likes its nature being dropped! Of course, in this case, that would just leave it to `concatenate` to deal with (which does not deal gracefully with subclasses, but that is for another issue).

Note: if there is agreement that it is fine to pass on subclasses, and that broadcasting is a good idea as well, then replace my above suggestion with `broadcast_arrays(*arrays, subok=True)`

jaimefrio replied:
The suggestion to broadcast `concatenate` arguments has been frowned upon on the list before, see [here](http://mail.scipy.org/pipermail/numpy-discussion/2014-August/071017.html). The idea seems to be that it is more likely to let errors pass unnoticed, than it is to actually be a useful functionality. I don't quite agree, but that change should be discussed on the list if we really wanted to pursue it.

mhvk replied:
For concatenate, I can see that broadcasting would be confusing, since along the axis one concatenates, different lengths are allowed. But for stacking, the argument would seem weaker.

Maybe the `broadcast=False` keyword you suggested in that thread? Although I guess it is really not that much more work to type `np.stack(np.broadcast_arrays(arrays), axis=...)`. 

Anyway, I don't feel strongly; the only reason to perhaps think a little more is that this is a new function, so this is good to try to get it right now.

shoyer replied:
I'm -0 on broadcasting automatically, though I know many existing functions in numpy do that. I do think it can hide errors and I don't think it's too hard to compose with `broadcast_arrays` explicitly if desired.

In any case, sure, I can use `np.asanyarray` instead... though it's not entirely clear to me if every ndarray subclass uses the same indexing rules (e.g., this would have broken on the old version of `pandas.Index`, when it used to be an ndarray subclass).

mhvk replied:
If subclasses mess with the indexing, I'd think it be up to the caller to ensure a proper `ndarray` is passed in -- it seems more logical to think of it as "I use a numpy function so what I put in should always look like an array" than "I use a numpy function so I every input always gets converted to an array".

mhvk replied:
Actually, an option may be to add a `subok` keyword argument, and then replace this line with:
```
arrays = [_nx.array(arr, copy=False, subok=subok) for arr in arrays]
```
(which should be slightly faster as it avoids the indirection of `asarray`)

It does mean adding to the documentation that how subclasses get treated in the end also depends on `concatenate`. In that respect, a default of `False` for subok may be reasonable.

njsmith replied:
Do the existing *_stack functions perform broadcasting?
On Feb 25, 2015 12:17 PM, "Marten van Kerkwijk" <notifications@github.com>
wrote:

> In numpy/core/shape_base.py
> <https://github.com/numpy/numpy/pull/5605#discussion_r25374756>:
>
> > +    >>> np.stack(arrays, axis=2).shape
> > +    (3, 4, 10)
> > +
> > +    >>> a = np.array([1, 2, 3])
> > +    >>> b = np.array([2, 3, 4])
> > +    >>> np.stack((a, b))
> > +    array([[1, 2, 3],
> > +           [2, 3, 4]])
> > +
> > +    >>> np.stack((a, b), axis=-1)
> > +    array([[1, 2],
> > +           [2, 3],
> > +           [3, 4]])
> > +
> > +    """
> > +    arrays = [_nx.asarray(arr) for arr in arrays]
>
> Actually, an option may be to add a subok keyword argument, and then
> replace this line with:
>
> arrays = [_nx.array(arr, copy=False, subok=subok) for arr in arrays]
>
> (which should be slightly faster as it avoids the indirection of asarray)
>
> It does mean adding to the documentation that how subclasses get treated
> in the end also depends on concatenate. In that respect, a default of
> False for subok may be reasonable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5605/files#r25374756>.
>

shoyer replied:
> Do the existing *_stack functions perform broadcasting?

Sort of, but poorly. They use the `atleast_Nd` functions, which will insert dimensions for input with `N - 1` or fewer dimensions. And the order in which those dimensions is not consistent with broadcasting, either (at least for `atleast_3d`):
```
In [4]: np.atleast_3d(np.zeros((2, 3))).shape
Out[4]: (2, 3, 1)

In [5]: np.atleast_3d(np.zeros((2))).shape
Out[5]: (1, 2, 1)
```

shoyer replied:
@mhvk I'm all for supporting ndarray subclasses when feasible, but the options we've come up with so far would either entail worse error reporting for non-compliant subclasses (using `asanyarray`) or additions to the function signature (adding `subok=False`) with no actual benefits, given the limitations of `concatenate`.

Are there any non-trivial ndarray subclasses that *would* benefit from these changes? I suppose we could argue that it is best to keep the door open to improvements in the future.

njsmith replied:
The problem with handling subclasses correctly in general is that it is
literally impossible. Subclasses might have completely arbitrary behavior,
so basically any behavior on the numpy side will be correct in some cases.
Should we preserve subclasses when doing X? Yes, that is the only correct
behavior (sometimes). Should we drop subclasses when doing X? Yes, that is
also required (other times). Should we return a different subclass?
Entirely possible. Use one specific set of indexing operations or function
calls instead of another purportedly equivalent approach? Definitely
necessary. Which specific set? Well, there are contradictory answers. Stand
on our heads? Can't rule it out...

Historically afaict the rule has been that we do whatever hack happens to
work for whichever subclass authors file a bug last...

What I'm saying is, I don't think we should waste too much energy worrying
about such questions because the best we can hope for is to choose who we
annoy :-/
On Feb 25, 2015 9:34 PM, "Stephan Hoyer" <notifications@github.com> wrote:

> In numpy/core/shape_base.py
> <https://github.com/numpy/numpy/pull/5605#discussion_r25405780>:
>
> > +    >>> np.stack(arrays, axis=2).shape
> > +    (3, 4, 10)
> > +
> > +    >>> a = np.array([1, 2, 3])
> > +    >>> b = np.array([2, 3, 4])
> > +    >>> np.stack((a, b))
> > +    array([[1, 2, 3],
> > +           [2, 3, 4]])
> > +
> > +    >>> np.stack((a, b), axis=-1)
> > +    array([[1, 2],
> > +           [2, 3],
> > +           [3, 4]])
> > +
> > +    """
> > +    arrays = [_nx.asarray(arr) for arr in arrays]
>
> @mhvk <https://github.com/mhvk> I'm all for supporting ndarray subclasses
> when feasible, but the options we've come up with so far would either
> entail worse error reporting for non-compliant subclasses (using
> asanyarray) or additions to the function signature (adding subok=False)
> with no actual benefits, given the limitations of concatenate.
>
> Are there any non-trivial ndarray subclasses that *would* benefit from
> these changes? I suppose we could argue that it is best to keep the door
> open to improvements in the future.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5605/files#r25405780>.
>

seberg replied:
There might be a way that works a bit more reliably if we do everything with arrays and call `__array_wrap__` (or similar) at the end. A few container functions do this (fI know of insert/delete). I think that might have less compatibility issues, though I doubt it will fix all troubles and not create troubles of its own....

shoyer replied:
@seberg  Yes, but unfortunately `__array_wrap__` won't work for many array-like objects if the shape of an array has changed. For example, consider `pandas.DataFrame`:
```
In [7]: df = pd.DataFrame(np.random.randn(2, 2))

In [8]: df.__array_wrap__(df.values)
Out[8]:
          0         1
0 -0.960405  1.736315
1 -1.277750  2.434818

In [9]: df.__array_wrap__(df.values[0])
ValueError: Shape of passed values is (1, 2), indices imply (2, 2)
```

There's no easy way out of this, aside from adding lots more hooks or multi-methods to push logic to the array-like objects/subclasses.

mhvk replied:
Agreed that the general problem is hard. However, `stack` is just a wrapper for `concatenate` and their would seem to be no reason for its behaviour to deviate from that of `concatenate` -- and hence for the inputs to be left as similar to what they are as possible.

But maybe a better argument is that this is what the other functions in `shape_base` do: the other `stack` functions use `at_least?d` and then call `concatenate`, and the `atleast_?d` functions use `asanyarray`. So, purely for consistency `asanyarray` is also called for.

mhvk replied:
p.s.  `insert` is indeed a good example of allowing subclasses to override standard behaviour -- it is on my list of to-do things (though fairly far down) to make `concatenate` somewhat similar such that its behaviour would be equivalent to creating a larger array with properties of the first one in the list, and then inserting the others. In that way, astropy's `Quantity` class could do unit verification and conversion if necessary.

shoyer replied:
OK, it turns out putting `np.matrix` objects in `stack` does still fail with a useful error message, despite the funny business with np.newaxis creating >2D matrix objects. So I no longer have any good reason to oppose using `asanyarray` here (though I will note that it makes more sense for functions like `vstack` and `hstack`, which *are* valid on matrix input). The latest commit makes this change.

I tried adding a test with `VerySimpleSubclass` (copied from `test_stride_tricks`) but it failed; `concatenate` appears to return base ndarray objects in that case. So for now, the subclass preserving behavior is mostly untested.

rgommers commented:
@shoyer did you see gh-5057, which also adds a ``stack`` function? A comparison with that PR may be useful (these PRs can't both be merged...).

shoyer commented:
@rgommers Good point. I did see that discussion on the mailing list; I did not realize that #5057 proposed using the same name.

I'll raise this point again on the mailing list. I think both types of functionality are useful, but this is a more fundamental type of "stacking" (it unifies the existing `*stack` functions) and fills a more obvious hole in NumPy's functionality (no N-dimensional way to stack ndarray objects).

shoyer commented:
I asked on the mailing list about the name `np.stack` last week: http://mail.scipy.org/pipermail/numpy-discussion/2015-March/072491.html

Based on @sotte (author of #5057) and @njsmith's comments, it seems like we would be OK to use stack here -- #5057 should probably use something else, such as `barray` or `block`.

mhvk commented:
Looks all OK.

njsmith commented:
Just looked over the api and I like it.

row_stack and column_stack are special in that they magically alternate
between stacking and concatenation depending on the dimensionality of the
input. But {h,v,d}stack are not so magical. Are they equivalent to
stack(..., axis=...)? (This is both a question about api consistency, and a
question about whether their implementation can be simplified.)
On Feb 25, 2015 2:38 AM, "Stephan Hoyer" <notifications@github.com> wrote:

> The motivation here is to present a uniform and N-dimensional interface
> for joining arrays along a new axis, similarly to how concatenate
> provides a uniform and N-dimensional interface for joining arrays along an
> existing axis.
> Background
>
> Currently, users can choose between hstack, vstack, column_stack and
> dstack, but none of these functions handle N-dimensional input. In my
> opinion, it's also difficult to keep track of the differences between these
> methods and to predict how they will handle input with different dimensions.
>
> In the past, my preferred approach has been to either construct the result
> array explicitly and use indexing for assignment, to or use np.array to
> stack along the first dimension and then use transpose (or a similar
> method) to reorder dimensions if necessary. This is pretty awkward.
>
> I brought this proposal up a few weeks on the numpy-discussion list:
> http://mail.scipy.org/pipermail/numpy-discussion/2015-February/072199.html
>
> I also received positive feedback on Twitter:
> https://twitter.com/shoyer/status/565937244599377920
> Implementation notes
>
> The one line summaries for concatenate and stack have been (re)written to
> mirror each other, and to make clear that the distinction between these
> functions is whether they join over an existing or new axis.
>
> In general, I've tweaked the documentation and docstrings with an eye
> toward pointing users to concatenate/stack/split as a fundamental set of
> basic array manipulation routines, and away from array_split/{h,v,d}split/
> {h,v,d,column_}stack
>
> I put this implementation in numpy.core.shape_base alongside hstack/vstack,
> but it appears that there is also a numpy.lib.shape_base module that
> contains another larger set of functions, including dstack. I'm not
> really sure where this belongs (or if it even matters).
>
> Finally, it might be a good idea to write a masked array version of stack.
> But I don't use masked arrays, so I'm not well motivated to do that.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/numpy/numpy/pull/5605
> Commit Summary
>
>    - ENH: add np.stack
>
> File Changes
>
>    - *M* doc/release/1.10.0-notes.rst
>    <https://github.com/numpy/numpy/pull/5605/files#diff-0> (3)
>    - *M* doc/source/reference/routines.array-manipulation.rst
>    <https://github.com/numpy/numpy/pull/5605/files#diff-1> (5)
>    - *M* numpy/add_newdocs.py
>    <https://github.com/numpy/numpy/pull/5605/files#diff-2> (3)
>    - *M* numpy/core/shape_base.py
>    <https://github.com/numpy/numpy/pull/5605/files#diff-3> (68)
>    - *M* numpy/core/tests/test_shape_base.py
>    <https://github.com/numpy/numpy/pull/5605/files#diff-4> (39)
>    - *M* numpy/lib/function_base.py
>    <https://github.com/numpy/numpy/pull/5605/files#diff-5> (2)
>    - *M* numpy/lib/index_tricks.py
>    <https://github.com/numpy/numpy/pull/5605/files#diff-6> (2)
>    - *M* numpy/lib/shape_base.py
>    <https://github.com/numpy/numpy/pull/5605/files#diff-7> (6)
>
> Patch Links:
>
>    - https://github.com/numpy/numpy/pull/5605.patch
>    - https://github.com/numpy/numpy/pull/5605.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5605>.
>

shoyer commented:
@njsmith I'm afraid `{h,v,d}stack` *are* so magical:
```
In [12]: x1 = np.zeros(1)

In [13]: x2 = np.zeros((1, 1))

In [14]: x3 = np.zeros((1, 1, 1))

In [16]: np.hstack([x1, x1]).shape
Out[16]: (2,)

In [17]: np.hstack([x2, x2]).shape
Out[17]: (1, 2)

In [18]: np.hstack([x3, x3]).shape
Out[18]: (1, 2, 1)

In [19]: np.vstack([x1, x1]).shape
Out[19]: (2, 1)

In [20]: np.vstack([x2, x2]).shape
Out[20]: (2, 1)

In [21]: np.vstack([x3, x3]).shape
Out[21]: (2, 1, 1)

In [22]: np.dstack([x1, x1]).shape
Out[22]: (1, 1, 2)

In [23]: np.dstack([x2, x2]).shape
Out[23]: (1, 1, 2)

In [24]: np.dstack([x3, x3]).shape
Out[24]: (1, 1, 2)
```

To summarize: if `array.ndim` < `stack_dim` (where `stack_dim` is 1 for hstack, 2 to vstack and 3 for dstack), then the `{h,v,d}stack` function stacks. Otherwise, it concatenates. There are currently no functions in the NumPy API that stack for arbitrary dimensional input -- that's why I made this PR.

From an implementation perspective, things are not so bad -- each of these methods just calls `np.atleast_*d` on all the inputs and then `np.concatenate` on the result. So in my opinion the API design problems are two fold:

1. The behavior of  `atleast_*d` is not very intuitive -- axes are inserted in somewhat random locations:

        In [45]: np.atleast_2d(np.zeros(2)).shape
        Out[45]: (1, 2)

        In [46]: np.atleast_3d(np.zeros(2)).shape
        Out[46]: (1, 2, 1)

        In [47]: np.atleast_3d(np.zeros((2, 2))).shape
        Out[47]: (2, 2, 1)

   If we were starting from scratch, I would pick the rule "always insert insert new dimensions at the start", which at least works like array broadcasting.
2. The axis that `{h,v,d}stack` concatenates along has no clear progression:

    - `hstack`: use `atleast_1d` and concatenates with `axis=1` (unless input is 1d, in which case `axis=0`)
    - `vstack`: use `atleast_2d` and concatenates with `axis=0`
    - `dstack`: use `atleast_3d` and concatenates with `axis=2`

Basically there is no way to understand these functions without reading the source code.

shoyer commented:
Ping! Does this need more work? I'd love to be able to merge this in time for numpy 1.10....

charris commented on numpy/core/shape_base.py:
```diff
@@ -275,3 +278,67 @@ def hstack(tup):
+
+    Parameters
+    ----------
+    arrays : sequence of ndarrays
```
The inputs are converted to arrays, so this should be "sequence of array_like"

charris commented on numpy/core/shape_base.py:
```diff
@@ -275,3 +278,67 @@ def hstack(tup):
+
+    """
+    arrays = [asanyarray(arr) for arr in arrays]
+    if not arrays:
```
`len(arrays) == 0` might be clearer.

shoyer replied:
IMO this is idiomatic way to write this in Python (but I don't care too much either way)

charris commented on numpy/core/shape_base.py:
```diff
@@ -275,3 +278,67 @@ def hstack(tup):
+    result_ndim = arrays[0].ndim + 1
+    if not -result_ndim <= axis < result_ndim:
+        raise IndexError('axis %r out of bounds [-%r, %r)'
+                         % (axis, result_ndim, result_ndim))
```
Probably put `%` on the preceding line.

mhvk replied:
Maybe use new-style formatting while one is at it?
```
IndexError('axis {0} out of bounds [-{1}, {1})'.format(axis, result_ndim))
```

charris commented on numpy/core/shape_base.py:
```diff
@@ -275,3 +278,67 @@ def hstack(tup):
+    sl = (slice(None),) * axis + (_nx.newaxis,)
+    try:
+        expanded_arrays = [arr[sl] for arr in arrays]
+    except IndexError:
```
Might be better to make this check explicit and move it up. You could do that by making a set of the shapes and checking that it has one member.
```
shapes = set(a.shape for a in arrays)
if len(shapes) == 0:
    raise...
```

shoyer replied:
yes, this is probably a good idea.

charris commented on numpy/core/shape_base.py:
```diff
@@ -275,3 +278,67 @@ def hstack(tup):
+
+def stack(arrays, axis=0):
+    """
+    Join a sequence of arrays along a new axis.
```
Need more explanation here. For instance in the examples, it isn't clear why `np.stack((a, b))` is not the same as `np.stack((a, b), axis-1)`.

shoyer replied:
I think the missing explanation is that the `axis` argument refers to the axis position in the *result* array, not in the input arrays.

charris replied:
Still needs a better explanation of where the new axis is located. It might also be better to follow the list insertion protocol.
```
In [7]: a = [1]*4

In [8]: a.insert(4, 2)

In [9]: a
Out[9]: [1, 1, 1, 1, 2]

In [10]: a = [1]*4

In [11]: a.insert(-1, 2)

In [12]: a
Out[12]: [1, 1, 1, 2, 1]
```

charris replied:
Although it is nice to indicate appending the axis with a simple `-1`. In the current case, an explanation of how negative axis values are handled would help.  For the list version you  could do something like
```
In [26]: a = ones((2,2,3))

In [27]: newshape = list(a.shape)

In [28]: newshape.insert(-1, 1)

In [29]: a.reshape(newshape).shape
Out[29]: (2, 2, 1, 3)
```

charris replied:
How about
```
def stack(arrays, axis=0):
    """
    Join a sequence of arrays along a new axis.

    The position of the new axis in the stacked result is specified by the
    `axis` parameter. For instance, if ``axis=-1`` the result will have
    been stacked along the last axis, and if ``axis=0`` it will have been
    stacked along the first axis.

    etc...
```

charris commented:
Does this deal with array scalers and empty arrays?

charris commented on numpy/core/shape_base.py:
```diff
@@ -275,3 +278,67 @@ def hstack(tup):
+           [3, 4]])
+
+    """
+    arrays = [asanyarray(arr) for arr in arrays]
```
What happens with mixed subtypes?

charris replied:
Could maybe check that all types are the same.

charris replied:
Could do that by checking that  `set(type(a) for a in arrays)` has one member.

shoyer replied:
What happens for subtypes is mostly dictated by the behavior of `np.concatenate`. I don't see much advantage in explicitly checking for consistent types here when none of the logic in this function relies on that.

mhvk replied:
I think the type checking should be left for `concatenate` (which does not currently do this all that well, but could be rewritten, e.g., using `insert` methods if present on the first member or so).

charris replied:
OK.

shoyer commented:
@charris It does seem to deal properly with array scalers and empty arrays -- I'll add some tests.

charris commented on numpy/core/shape_base.py:
```diff
@@ -275,3 +278,67 @@ def hstack(tup):
+                         % (axis, result_ndim, result_ndim))
+    if axis < 0:
+        axis += result_ndim
+    sl = (slice(None),) * axis + (_nx.newaxis,)
```
An alternative method, once you have the shape of the arrays, is
```
newshape = shape[:axis] + (1,) + shape[axis:]
expanded_arrays = [a.reshape(shape) for a in arrays]
```

charris replied:
Or, getting rid of `expanded_arrays`
```
_nx.concatenate([a.reshape(newshape) for a in arrays], axis=axis)
```

shoyer replied:
I like using slicing for this operation rather than reshape because I *know* that slicing will always using a view rather than a copy. Though I suppose `reshape` is probably also safe when used in this way.

shoyer commented:
My latest commit includes changes in response to @charris's review (thanks!)

charris commented:
@shoyer Some more inprovement in the summary part of the docstring is needed, then this can go in.

shoyer commented:
@charris I added your suggested words to the docstring and squashed my commits. I'm still waiting for Travis to run its tests, but they did pass (except for `USE_BENTO=1`) on my fork:
https://travis-ci.org/shoyer/numpy/builds/62164157

charris commented on numpy/core/shape_base.py:
```diff
@@ -275,3 +278,73 @@ def hstack(tup):
+    """
+    Join a sequence of arrays along a new axis.
+
+    .. versionadded:: 1.10.0
```
`.. versionadded:: 1.10.0` and its preceding blank line should come at the end of the summary ;)

shoyer replied:
Fixed.

charris commented:
Looks good. One more nitpick and it's done...

charris commented:
Great. Thanks Stephan.
